### PR TITLE
support delayed scaling of weight in float8 all-gather

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #312
* #311
* #310
* #309

Summary:

Adds support for delayed scaling in FSDP2 float8 all-gather. In detail:
1. add `WeightWithDelayedFloat8CastTensor`, note that we don't reuse
   code with the dynamic version because I'd rather not deal with
   plumbing optional tensors through dynamo. We can try that in a
   separate PR later.
2. wire `Float8Linear` to use (1)
3. add weight amax syncing back, since we need it for float8 all-gather
4. add test coverage for eager mode numerics

Next up (in separate PRs) will be training run validation for numerics, and
taking a look at performance.

Test Plan:

```
./test/test_everything.sh
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D59685258](https://our.internmc.facebook.com/intern/diff/D59685258)

### CODE REVIEW

**Comment by weifengpy** [comment]
> I'd rather not deal with
plumbing optional tensors through dynamo

what are the optional tensors ?

File: float8_experimental/float8_linear.py [hunk 1]
```diff
@@ -316,25 +319,28 @@ def cast_w_to_float8(
...
+                    reduce_amax=False,
+                )
+
+                w_fp8 = Float8Tensor.to_float8(
+                    w,
+                    self.fp8_scale_w,
+                    e4m3_dtype,
+                    self.fp8_amax_w,
+                    self.forward_config,
+                )
```
Comment (drisspg) [inline float8_experimental/float8_linear.py:343]: There is a todo below which is a lil confusing

File: float8_experimental/float8_linear_utils.py [hunk 2]
```diff
@@ -291,11 +291,10 @@ def inner_func():
         ), "Mismatched lengths of amax tensors."
 
         if dist.is_initialized():
-            # Combine all the amax tensors into one tensor and reduce it
-            # Note: do not reduce the weight values, because FSDP already ensures
-            # the weight values on all ranks are the same after all-gather.
             all_amax_tensors = torch.cat(
-                fp8_amax_x_tensor_list + fp8_amax_dL_dY_tensor_list
+                fp8_amax_x_tensor_list
+                + fp8_amax_w_tensor_list
```
Comment (drisspg) [inline float8_experimental/float8_linear_utils.py:296]: should we only do this if we are using fp8 all gather ?

File: float8_experimental/fsdp_utils.py [hunk 3]
```diff
@@ -110,3 +112,181 @@ def fsdp_post_all_gather(
             out._scale = scale
             return
         return Float8Tensor(data, scale, param_dtype, self._mm_config), (data,)
+
+
+class WeightWithDelayedFloat8CastTensor(torch.Tensor):
```
Comment (drisspg) [inline float8_experimental/fsdp_utils.py:117]: [no change needed] I wish there was a way to share some more code with the dynamic linear version

File: float8_experimental/fsdp_utils.py [hunk 4]
```diff
@@ -110,3 +112,181 @@ def fsdp_post_all_gather(
...
+
+    def __tensor_flatten__(self):
+        return (
+            [
+                "_tensor",
+                "_amax_buffer",
+                "_amax_history_buffer",
+                "_scale_buffer",
+            ],
+            self._mm_config,
```
Comment (drisspg) [inline float8_experimental/fsdp_utils.py:225]: nit: I think keeping the non tensors attributes in a dict is a lil nicer: https://github.com/pytorch-labs/float8_experimental/blob/13d319858d8c2e9e0bef0e509ac753b84ac08fa6/float8_experimental/float8_tensor.py#L233-L238

File: float8_experimental/fsdp_utils.py [hunk 5]
```diff
@@ -110,3 +112,181 @@ def fsdp_post_all_gather(
...
+            inner_tensors["_amax_history_buffer"],
+            inner_tensors["_scale_buffer"],
+            mm_config,
+            is_amax_initialized,
+        )
+
+    def __repr__(self):
+        return f"WeightWithDelayedFloat8CastTensor(tensor={self._tensor}, amax_buffer={self._amax_buffer}, scale_buffer={self._scale_buffer}, mm_config={self._mm_config})"
+
+    def fsdp_pre_all_gather(self, mesh):
```
Comment (drisspg) [inline float8_experimental/fsdp_utils.py:244]: ill let @weifengpy confirm this portion

File: float8_experimental/float8_linear.py (hunk 1 above)
Comment (vkuzo) [reply float8_experimental/float8_linear.py:343]: thanks, will remove the TODO

File: float8_experimental/float8_linear_utils.py (hunk 2 above)
Comment (vkuzo) [reply float8_experimental/float8_linear_utils.py:296]: that could make sense, I'd love to see the data to see if this is going to matter for performance.  Focusing on numerics for now, was hoping for performance be tackled in future PRs.

File: float8_experimental/fsdp_utils.py (hunk 3 above)
Comment (vkuzo) [reply float8_experimental/fsdp_utils.py:117]: yeah, me too.  Looking at the code below, really the only code which would be shared is `fsdp_post_all_gather`, everything else would have to have if/else branches for delayed vs dynamic

File: float8_experimental/fsdp_utils.py (hunk 4 above)
Comment (vkuzo) [reply float8_experimental/fsdp_utils.py:225]: makes sense, will do!

File: float8_experimental/fsdp_utils.py [hunk 6]
```diff
@@ -110,3 +112,182 @@ def fsdp_post_all_gather(
...
+        self._mm_config = mm_config
+
+        # Note: is_amax_initialized is not a buffer to avoid data dependent
+        # control flow visible to dynamo
+        # TODO(future PR): add serialization for this flag
+        self.is_amax_initialized = is_amax_initialized
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs=None):
+        if func == torch.ops.aten.detach.default:
```
Comment (bdhirsh) [inline float8_experimental/fsdp_utils.py:163]: mostly just a nit, but any reason to special-case detach here? Alternatively, you could set it up so that every view ops automatiomatically propagates subclass-ness in the same way

Comment (awgu) [reply float8_experimental/fsdp_utils.py:163]: If this is something I wrote, I think it was just something I saw in some other subclasses. Having every view up propagate subclass-ness in the same way sounds good to me.

File: float8_experimental/fsdp_utils.py (hunk 5 above)
Comment (weifengpy) [reply float8_experimental/fsdp_utils.py:245]: confirming that fsdp part looks good

**Review by weifengpy — APPROVED** [review]
stamping for the fsdp part

document 2 open questions (not blocker for this PR)
* should we merge `WeightWithDelayedFloat8CastTensor` and `WeightWithDynamicFloat8CastTensor` into one class and add if-else to unify logic around `__torch_dispatch__`, `fsdp_pre_all_gather`/`fsdp_post_all_gather`. we unifed `Float8Linear` already
* compare perfs between `sync_float8_amax_and_scale_history` and `precompute_float8_dynamic_scale_for_fsdp`. If they are similar, people would not need to worry about numeric problem from delayed scaling

**Comment by vkuzo** [comment]
> should we merge WeightWithDelayedFloat8CastTensor and WeightWithDynamicFloat8CastTensor into one class and add if-else to unify logic around __torch_dispatch__, fsdp_pre_all_gather/fsdp_post_all_gather. we unifed Float8Linear already

I'm open if someone is interested in doing that in a follow-up PR.  I'm not sure it will be better than what we have now though.  Note that `Float8Linear` was unified to allow for finer grained configuration of scaling (per-tensor instead of per-module), that benefit is not on the table here.

> compare perfs between sync_float8_amax_and_scale_history and precompute_float8_dynamic_scale_for_fsdp. If they are similar, people would not need to worry about numeric problem from delayed scaling

yes, that would be great!  I think we can do this in follow-up PRs.  Note that delayed scaling is theoretically faster than dynamic scaling (less memory reads), but performance is not optimized across the stack yet.  I think it's good to have options and allow people to optimize different settings in parallel.  Eventually if there is clear data that only one of these is needed, we can delete the not-needed ones.

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59685258).

File: float8_experimental/fsdp_utils.py [hunk 7]
```diff
@@ -189,3 +189,182 @@ def fsdp_post_all_gather(
...
+            metadata["mm_config"],
+            metadata["is_amax_initialized"],
+        )
+
+    def __repr__(self):
+        return f"WeightWithDelayedFloat8CastTensor(tensor={self._tensor}, amax_buffer={self._amax_buffer}, scale_buffer={self._scale_buffer}, mm_config={self._mm_config})"
+
+    def fsdp_pre_all_gather(self, mesh):
+        # initialize if needed
+        # TODO(before land): ensure settings are consistent between Float8Linear and here
```
Comment (awgu) [inline float8_experimental/fsdp_utils.py:324]: do we still need to resolve this?

File: float8_experimental/fsdp_utils.py [hunk 8]
```diff
@@ -189,3 +189,182 @@ def fsdp_post_all_gather(
...
+            from float8_experimental.float8_linear import (
+                _maybe_initialize_amaxes_scales_for_float8_cast,
+            )
+
+            _maybe_initialize_amaxes_scales_for_float8_cast(
+                self._tensor,
+                self._amax_buffer,
+                self._amax_history_buffer,
+                self._scale_buffer,
+                "max",  # TODO(before land): read this from parent
```
Comment (awgu) [inline float8_experimental/fsdp_utils.py:335]: ditto
